### ISSUE TITLE
Summing elements of an array - 'Computed' hangs.

### ISSUE THREAD

**Issue by mungewell** [construct/construct#893]
I am trying to define (reverse engineer) a format which has an array of sample quantities, for each of 12 elements. I would like to sum this array so that I can count the total number of samples... but the 'Computed' function just hangs indefinately.
```
Sample = Struct(
    "length" / Int24ul,
    Const(b"\x7d\x53\x4d\x50\x00"),
)

Block = Struct(
    "counts" / Array(12, Byte),
    "total" / Computed(sum([51,2,1])),      # Works
    #"total" / Computed(sum(this.counts)),  # Hangs indefinately
    "samples" / Array(this.total, Sample),
)
```

Later (in normal Python) I can sum the array...
```
    if data: 
        config = DFU.parse(data)
        if options.dump:
            #print(config)
            print("total samples:", sum(config["counts"]))
```

Any suggestions about how to code a Struct for this? Thanks,
Simon.

**Comment by arekbulski** [comment]
Computer parameter is (must be) a callable object. The builtin sum does not return a callable object, it returns a number.

Should be: `"total" / Computed(lambda this: sum(this.counts)),`

**Comment by mungewell** [comment]
Thanks, that appears to work... now to figure out how to sum items enclosed within an array. Will see what I can figure and ask if confused.

**Comment by arekbulski** [comment]
Maybe use an Adapter?

**Comment by mungewell** [comment]
Happy to report it was actually pretty easy using `this._index`, Python is awesome. I was able to build an array of data/Bytes blocks where their length was controlled by a different (previously read/parsed) array.

```
Data = Struct(
    "data" / Bytes(lambda this: this._.samples[this._index].bytes),
)
```

Full code is here if anyone is interested:
https://github.com/mungewell/uno-drum-utils/blob/main/decode_sound_packs.py#L51

I have also used an adapter in an odd way to read the number of (12bit) samples in a block which the proprioratory format stores, and magically convert this to the number of bytes required to store them.
